### PR TITLE
Porting the environment switcher into a common location

### PR DESCRIPTION
**Pull request by 1egoman**
We've historically had a lot of duplicates of the environment switcher throughout a bunch of different projects. This is the start of an effort to begin to standardize on one canonical environment switcher across applications.

### CODE REVIEW

File: src/environment-switcher/index.tsx [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+export const EnvironmentSwitcher: EnvironmentSwitcherComponentType = ({
+  localEnvironmentControls,
+  localEnvironmentDefaultValue,
+  customEnvironmentControls,
+  customEnvironmentDefaultValue,
+
+  globalControls,
+  globalSettings,
+}) => {
+  // FIXME: fix modal
```
Comment (1egoman) [inline src/environment-switcher/index.tsx:143]: Ah, oops, I still need to fix this!

File: src/environment-switcher/index.tsx [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+  configuration: C /* version of the configuration used by the app */;
+};
+
+
+type EnvironmentSwitcherProps<
+  L extends object,
+  C extends object,
+  G extends object,
+> = {
+  // Settings shown in "More options" panel
```
Comment (1egoman) [inline src/environment-switcher/index.tsx:105]: @jshanley This is where some of the complicated types I was referring to are located.

Comment (guscost) [reply src/environment-switcher/index.tsx:105]: I'm on the fence about structuring the settings this way, it seems like "Settings" could contain "Environment", but all of this conditional relevance makes me question it and want to agree with your approach. Is it going to be more common for a future setting to be "global" or "particular to one environment"? Not sure.

Anyway, the implementation looks good in general.

File: src/environment-switcher/index.tsx [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+    },
+  ];
+
+  // NOTE: This needs to be inside here so this this type can vary based on the types of some of the props.
+  type EnvironmentDefinition = (
+    | { type: Environment.PRODUCTION} & typeof globalSettings
+    | { type: Environment.STAGING } & typeof globalSettings
+    | { type: Environment.LOCAL } & typeof localEnvironmentDefaultValue & typeof globalSettings
+    | { type: Environment.CUSTOM } & typeof customEnvironmentDefaultValue & typeof globalSettings
+  );
```
Comment (guscost) [inline src/environment-switcher/index.tsx:185]: 🤷

**Review by guscost — APPROVED** [review]
Looks great in general.

File: src/environment-switcher/index.tsx (hunk 2 above)
Comment (1egoman) [reply src/environment-switcher/index.tsx:105]: > Is it going to be more common for a future setting to be "global" or "particular to one environment"? Not sure.

I think equally common, or more accurately, it depends on the setting.

**Pull request merged**